engine/consensus/matching/core.go

```diff
@@ -562,222 +559,244 @@ func (c *Core) CheckSealing() error {
-	}
-
-	nextUnsealedHeight := lastSealed.Height + 1
-	nextUnsealed, err := state.AtHeight(nextUnsealedHeight).Head()
-	if errors.Is(err, storage.ErrNotFound) {
-		// next unsealed block has not been finalized yet.
-		return flow.ZeroID, false, nil
-	}
-	if err != nil {
-		return flow.ZeroID, false, fmt.Errorf("could not get block at heigh:%v, %w", nextUnsealed, err)
-	}
-	return nextUnsealed.ID(), true, nil
-}
-
 // sealableResults returns the IncorporatedResults from the mempool that have
 // collected enough approvals on a per-chunk basis, as defined by the matchChunk
 // function. It also filters out results that have an incorrect sub-graph.
 // It specifically returns the information for the next unsealed results which will
 // be useful for debugging the potential sealing halt issue
-func (c *Core) sealableResults() ([]*flow.IncorporatedResult, nextUnsealedResults, error) {
-	var results []*flow.IncorporatedResult
+func (c *Core) sealableResults() ([]*flow.IncorporatedResult, *sealingtracker.SealingTracker, error) {
+	//tracking functionality to collection information about the sealing status.
+	sealingTracker := sealingtracker.NewSealingTracker(c.state)
 
 	lastFinalized, err := c.state.Final().Head()
 	if err != nil {
 		return nil, nil, fmt.Errorf("failed to get last finalized block: %w", err)
 	}
 
-	nextUnsealed, nextUnsealedIsFinalized, err := nextUnsealedID(c.state)
-	if err != nil {
-		return nil, nil, fmt.Errorf("failed to get the next unsealed block id: %w", err)
-	}
-
-	nextUnsealeds := make([]*nextUnsealedResult, 0)
-	// go through the results mempool and check which ones we have collected
-	// enough approvals for
+	// go through the results mempool and check which ones we can construct a candidate seal for
+	var results []*flow.IncorporatedResult
 	for _, incorporatedResult := range c.incorporatedResults.All() {
-		// not finding the block header for an incorporated result is a fatal
-		// implementation bug, as we only add results to the IncorporatedResults
-		// mempool, where _both_ the block that incorporates the result as well
-		// as the block the result pertains to are known
-		block, err := c.headersDB.ByBlockID(incorporatedResult.Result.BlockID)
-		if err != nil {
-			return nil, nil, fmt.Errorf("could not retrieve block: %w", err)
-		}
-
-		// At this point we can be sure that all needed checks on validity of ER
-		// were executed prior to this point, since we perform validation of every ER
-		// before adding it into mempool. Hence, the mempool can contain only valid entries.
-
-		// the chunk assigment is based on the first block in its fork which
-		// contains a receipt that commits to this result.
-		assignment, err := c.assigner.Assign(incorporatedResult.Result, incorporatedResult.IncorporatedBlockID)
+		// Can we seal following the happy-path protocol?
+		sealingStatus, err := c.hasEnoughApprovals(incorporatedResult, sealingTracker)
```

Comment:
⚠️ This behavior is different from what I expected. 

What I expected was that: 
1. we will require multiple receipts to produce a seal
2. in other words, we will never produce a seal if there is less two receipts.

Now the new behavior is that:
1. we will require multiple receipts _and_ either enough approvals or emergency sealing to produce a seal
2. in other words, we won't produce a seal even if we've received multiple receipts, but haven't received enough approvals yet, and emergency sealing has not been triggered.

I'm not comfortable to have the sealing logic requiring enough approvals, because we still have [one unsolved issue](https://github.com/dapperlabs/flow-go/issues/5195), which is verification node can't resume the verification on results which are produced when it's off. So I'm afraid the emergency sealing will often be triggered, and makes the sealing sometimes fast and sometimes slow, making it hard to debug.

I would suggest to keep the original plan to just seal with multiple receipts until we fix [the sealing liveness issue](https://github.com/dapperlabs/flow-go/issues/5195)

Reply:
You are right that the logic seals under the condition
> we will require multiple receipts _and_ either enough approvals or emergency sealing to produce a seal

This is also stated here in the code comment: 
https://github.com/onflow/flow-go/blob/34a14af217846c01cae29301a1c4837ea4137529/engine/consensus/matching/core.go#L605-L612

Essentially, the consensus node already implements the general case, where we could require some approvals to seal (or wait until emergency sealing kicks in). With the parametrization `DefaultRequiredApprovalsForSealConstruction = 0`, the function `hasEnoughApprovals` always returns `true` (for _any_ `incorporatedResult`). Therefore, `(i)` is _always_ true and the sealing condition 
```
 //  (i) the incorporatedResult must qualify for happy path sealing 
 //      or qualify for emergency sealing 
 // AND 
 // (ii) there must be at least 2 receipts from _different_ ENs 
 //      committing to the result 
```
reduces to 
```
there must be at least 2 receipts from _different_ ENs  committing to the result 
``
which is exactly what you expected. 

### Note: 
From my perspective, **the current logic _should_ be the same as what we are currently running on Mainnet** (PR https://github.com/onflow/flow-go/pull/444):
https://github.com/onflow/flow-go/blob/c4ea8726dff06c912a3e6aeee2128b526421a135/engine/consensus/matching/engine.go#L812-L817

Reply:
I have added the following explicit shortcut in `hasEnoughApprovals` 
https://github.com/onflow/flow-go/blob/4bdb072bd510b3b356e2ae39475e30996aeea10a/engine/consensus/matching/core.go#L646-L649
* which hopefully makes the logic clearer
* it shortcuts some work of determining assignments if we don't require any approvals to seal